dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/DefaultTrackerNotificationWebHookService.java

```diff
@@ -0,0 +1,157 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.google.common.collect.Lists;
+
+/**
+ * @author Zubair Asghar
+ */
+@Slf4j
+@RequiredArgsConstructor
+@Service( "org.hisp.dhis.program.notification.TrackerNotificationWebHookService" )
+public class DefaultTrackerNotificationWebHookService implements TrackerNotificationWebHookService
+{
+    @NonNull
+    private final ProgramInstanceStore programInstanceStore;
+
+    @NonNull
+    private final ProgramStageInstanceStore programStageInstanceStore;
+
+    @Nonnull
+    private final RestTemplate restTemplate;
+
+    @Nonnull
+    private final RenderService renderService;
+
+    @Override
+    @Transactional
+    public void handleEnrollment( String pi )
+    {
+        ProgramInstance instance = programInstanceStore.getByUid( pi );
+
+        List<ProgramNotificationTemplate> templates = instance.getProgram().getNotificationTemplates()
+            .stream()
+            .filter( t -> ProgramNotificationRecipient.WEB_HOOK == t.getNotificationRecipient() )
+            .collect( Collectors.toList() );
+
+        Map<String, String> payload = new HashMap<>();
+        ProgramNotificationMessageRenderer.VARIABLE_RESOLVERS
+            .forEach( ( key, value ) -> payload.put( key.name(), value.apply( instance ) ) );
+        sendPost( templates, renderService.toJsonAsString( payload ) );
+    }
+
+    @Override
+    @Transactional
+    public void handleEvent( String psi )
+    {
+        ProgramStageInstance instance = programStageInstanceStore.getByUid( psi );
+
+        List<ProgramNotificationTemplate> templates = instance.getProgramStage().getNotificationTemplates()
+            .stream()
+            .filter( t -> t.getNotificationRecipient() == ProgramNotificationRecipient.WEB_HOOK )
+            .collect( Collectors.toList() );
+
+        Map<String, String> payload = new HashMap<>();
+        ProgramStageNotificationMessageRenderer.VARIABLE_RESOLVERS
+            .forEach( ( key, value ) -> payload.put( key.name(), value.apply( instance ) ) );
+        sendPost( templates, renderService.toJsonAsString( payload ) );
+    }
+
+    private void sendPost( List<ProgramNotificationTemplate> templates, String payload )
+    {
+        ResponseEntity<String> responseEntity = null;
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.put( "Content-type", Lists.newArrayList( "application/json" ) );
+
+        HttpEntity<String> httpEntity = new HttpEntity<>( payload, httpHeaders );
+
+        for ( ProgramNotificationTemplate t : templates )
+        {
+            URI uri = UriComponentsBuilder.fromHttpUrl( t.getMessageTemplate() ).build().encode().toUri();
```

Comment:
Should we for safety check that the template don't include any extra whitespace of newlines? maybe do some basic trimming?

Reply:
We can also do validation early that we have a valid URL here, `ValidationUtils`.`urlIsValid`, with a corresponding exception with error code.

Reply:
resolved